haskell/private/actions/info.bzl

```diff
@@ -0,0 +1,181 @@
+    "-package=directory",
+]
+
+def _filter_package_env(flags):
+    # Strips out -package-env from the command-line flags.  Consumers of these output
+    # groups will be responsible for setting the right GHC flags themselves,
+    # based on the fields of haskell.LibraryInfo.
+    result = []
+    for i in flags:
+        if not flags:
+            break
+        if flags[0] == "-package-env":
+            flags = flags[2:]
+        else:
+            result.append(flags[0])
+            flags = flags[1:]
+    return result
+
+def _write_haskell_compile_info(
+        workspace_name,
+        name,
+        hs,
+        c,
+        cc_libs,
+        runfiles):
+    return write_proto_file(
+        output_name = name,
+        hs = hs,
+        proto_type = "haskell.CompileInfo",
+        content = struct(
+            # Calling to_list on c.source_files shouldn't be a performance penalty.
+            # Despite being a depset, it only contains sources for the current rule.
+            source_files = [f.path for f in c.source_files.to_list()],
+            # TODO: currently, this will duplicate the common, target-independent options for
+            # each build target.  We should instead move them into GhcConfig.common_options.
+            options = _filter_package_env(c.compile_flags) + _COMMON_REPL_OPTIONS,
+            transitive_cc_shared_libs = [lib.path for lib in cc_libs],
+            # Follows the new runfiles tree organization of:
+            # https://github.com/bazelbuild/bazel/wiki/Updating-the-runfiles-tree-structure
+            runfiles = [
+                struct(
+                    full_path = f.path,
+                    short_path = paths.join(
+                        f.owner.workspace_name or workspace_name, f.short_path))
+                for f in runfiles.to_list()
+            ],
+        ),
+    )
+
+
+def library_info_output_groups(
+        name,
+        hs,
+        hs_info,
+        lib_info):
+    """Output groups for depending on a Haskell target.
+
+    Args:
+        name: A string; the name of the current target.
+        hs: The Haskell context.
+        hs_info: A HaskellInfo provider.
+        lib_info: A HaskellLibraryInfo provider.
+
+    Returns:
+      A dict whose keys are output groups and values are depsets of Files.
+    """
+    proto_file = write_proto_file(
+        hs = hs,
+        output_name = name + ".HaskellLibrary",
+        proto_type = "haskell.LibraryInfo",
+        content = struct(
+            # TODO(google/hrepl#4): currently, we only expose the immediate dependencies.
+            transitive_package_ids = [lib_info.package_id],
+            transitive_package_dbs =
+                [db.dirname for db in hs_info.package_databases.to_list()]))
+    return {
+        "haskell_transitive_deps": depset(
+          transitive = [
+                hs_info.package_databases,
+                hs_info.interface_dirs,
+                hs_info.dynamic_libraries,
+            ]),
```

Comment:
C library dependencies are triggered through `compile_info_output_groups()`; specifically, the `haskell_cdeps_shared_lib` output group.  It should contain transitive C/C++ libraries for all the targets that are being compiled.

In other words,

Reply:
> In other words,

Did the comment get cropped?

Reply:
Finishing the comment:

In other words, `CompileInfo` provides all files/information about interpreting/running the targets passed on the command-line, and `LibraryInfo` provides files/information about compiling their dependencies.  Since C++ libraries are used at runtime, they're attached to the former.

Reply:
Understood, thanks for explaining! Since `CcInfo` carries transitive C library dependencies this will also contain C libraries introduced by other Haskell dependencies. Makes sense.